### ISSUE TITLE
Another Diversity-Promoting Objective Function for Neural Dialogue Generation

### ISSUE THREAD

**Issue by noowad93** [noowad/paper-summaries#34]
## 논문 정보
AAAI 2019
Another Diversity-Promoting Objective Function for Neural Dialogue Generation
Ryo Nakamura, Katsuhito Sudoh, Koichiro Yoshino, Satoshi Nakamura
NAIST
## 개요
뉴럴 대화 모델이 뻔한을 주로 생성하는 가장 큰 이유는 Softmax Cross-Entropy loss + 최대우도추정에 기인한다. 본 연구에서는 새로운 목적 함수인 Inverse Token Frequency (ITF) 로스를 제안하여 이러한 문제를 해결하려는 논문이다.
## 선행 연구보다 발전된 점
이전부터 MMI와 같은 목적 함수를 이용하거나, GAN, VAE 등의 생성 모델을 이용하는 등 최대우도추정의 목적 함수를 수정하여 뻔한 답변을 해결하려는 연구들이 연구되어 왔다. 본 연구에서 제안하는 ITF 로스는 이러한 수법들보다 좀 더 간단명료하면서도 안정되게 훈련이 가능하다는 장점이 있다.
## 어떠한 수법을 사용하였는가?
본 연구에서는 seq2seq를 이용한 생성 방식의 모델을 가정한다. 이 모델은 타겟 토큰 별로 로스를 계산한다.
일반적인 소프트맥스 크로스엔트로피 로스는 아래의 식과 같이 seq2seq 모델을 훈련한다.
<img width="385" alt="스크린샷 2019-05-31 오후 3 07 04" src="https://user-images.githubusercontent.com/49383323/58685176-c6d9ea00-83b5-11e9-913f-d6d1bc3311ce.png">
위 식은 모든 토큰 클래스를 동등하게 간주한다. 따라서 고빈도 코튼의 생성 확률이 매우 커지고, 저빈도 토큰의 생성 확률이 매우 작아진다는 단점이 있다. 반면에, 본 연구에서는 아래와 같은 ITF 로스를 제안한다.
<img width="366" alt="스크린샷 2019-05-31 오후 3 08 39" src="https://user-images.githubusercontent.com/49383323/58685242-fc7ed300-83b5-11e9-8d49-dba4ebb30dbf.png">
즉, SCE 로스에서 각 토큰에 빈도별로 가중치를 매긴 값이다. lambda는 하이퍼파라미터이고, 이 값이 0일때는 SCE 로스와 동일하다. 단, START나 END와 같은 스페셜 토큰들도 다른 토큰들과 동일하게 간주하여, 이러한 토큰들이 매우 낮은 가중치를 갖도록 한다. 
<img width="450" alt="스크린샷 2019-05-31 오후 3 11 08" src="https://user-images.githubusercontent.com/49383323/58685335-55e70200-83b6-11e9-9e73-2adfb401191c.png">
PyTorch 버전으로 구현한다면 이런 식으로

## 검증
실험은 언어와 도메인이 다른 3가지 데이터셋으로 진행했다.
베이스라인 모델들로는
* seq2seq
* seq2seq+attention
* seq2seq+MMI objective
이 모델은 Li 2016a에서 제안된 방법으로, 
<img width="411" alt="스크린샷 2019-05-31 오후 3 16 44" src="https://user-images.githubusercontent.com/49383323/58685557-1e2c8a00-83b7-11e9-9a5b-d58bd8787489.png">
위와 같이 생성 문장의 생성확률 (언어 모델)의 대수 우도 값을 벌칙 항으로 추가한 objective를 이용한 모델이다.
(언어 모델적으로 확률이 높다는 의미는 빈도가 높은 문장이라는 의미이므로)
여기서는 학습시에 이 objective를 이용해도 diversity의 개선이 없었으므로, 훈련할 때는 단순 MLE, inference할 때는 별도의 언어 모델을 이용하여 위의 로스를 이용했다고 함.
* MemN2N
 대화 생성 모델에서는 문맥상의 consistency를 위해서 멀티턴 발화를 dialogue history로 이용하기 위해 Memory Network나 HRED 등을 이용하여 모델링하는 경우가 많음. 여기서는 멀티턴 대화 이력을 메모리로 활용하는 방식으로 메모리 네트워크를 이용함.

본 연구에서는 생성 문장의 퀄리티를 측정하기 위해 BLEU를 이용했고, diversity를 측정하기 위해 DIST를 이용했다.
DIST-n은 생성 문장의 n-grams 들 중 distinct한 n-grams의 비율을 측정하는 척도이다.

결과는 아래와 같다. 
<img width="486" alt="스크린샷 2019-05-31 오후 3 44 20" src="https://user-images.githubusercontent.com/49383323/58686894-140c8a80-83bb-11e9-934d-16ce08eff2e5.png">
<img width="479" alt="스크린샷 2019-05-31 오후 3 44 24" src="https://user-images.githubusercontent.com/49383323/58686900-15d64e00-83bb-11e9-9dec-76a87f13b428.png">

이 외에도, 학습할 때는 단순 MLE, 추론시에만 ITF를 적용하는 식의 방식을 적용해봤을 때에도 diversity가 증가하는 것을 보임.
<img width="476" alt="스크린샷 2019-05-31 오후 3 47 57" src="https://user-images.githubusercontent.com/49383323/58687057-7cf40280-83bb-11e9-80a2-08d502cfca4d.png">

기타 생성 결과 등은 논문을 참조
## 개인적인 생각
논문에서 지적한 바와 같이 일상 대화에서 likelihood가 제일 높은 답변이 대화 관점에서의 좋은 답변이라고는 할 수 없기 때문에, 목적 함수를 변경하는 방식의 접근은 앞으로 계속 연구되어야할 방향이라 생각함.
다만, 빈도별로 가중치를 매긴 방법은 일반적으로 imbalanced classes 문제를 해결할 때 적용하는 class weight 방식과 유사한 부분이 있기 때문에 novelty가 큰 논문이라고는 생각하지 않음.
